[14.0] edi framework misc improvements

simahawk opened the pull request:
See atomic commits for details.

simahawk committed:
edi_backend: update docstring

simahawk committed:
edi_backend: fix var name

simahawk committed:
edi: exc record get file content as b64 optionally

simahawk committed:
edi: improve exchange record UI

simahawk committed:
edi: improve backend view extension

simahawk committed:
edi: allow job channel conf by type

simahawk committed:
edi: specific method to get type settings

simahawk commented:
@etobella would you have time for a review?

etobella commented on edi_oca/models/edi_backend.py:
```diff
@@ -330,7 +337,7 @@ def _check_output_exchange_sync(self, skip_send=False):
         )
         for rec in new_records:
-            self.with_delay().exchange_generate(rec)
+            self.with_delay(self._get_job_delay_params(rec)).exchange_generate(rec)
```
self.with_delay(**self._get_job_delay_params(rec)).exchange_generate(rec)

etobella commented on edi_oca/models/edi_backend.py:
```diff
@@ -343,7 +350,7 @@ def _check_output_exchange_sync(self, skip_send=False):
         for rec in pending_records:
             if rec.edi_exchange_state == "output_pending":
-                self.with_delay().exchange_send(rec)
+                self.with_delay(self._get_job_delay_params(rec)).exchange_send(rec)
```
self.with_delay(**self._get_job_delay_params(rec)).exchange_send(rec)

etobella commented on edi_oca/models/edi_backend.py:
```diff
@@ -521,7 +528,7 @@ def _check_input_exchange_sync(self, **kw):
         )
         for rec in pending_records:
-            self.with_delay().exchange_receive(rec)
+            self.with_delay(self._get_job_delay_params(rec)).exchange_receive(rec)
```
self.with_delay(**self._get_job_delay_params(rec)).exchange_receive(rec)

etobella commented on edi_oca/models/edi_backend.py:
```diff
@@ -559,7 +566,9 @@ def _exchange_check_ack_needed(self, pending_records):
         )
         for rec in ack_pending_records:
-            self.with_delay().exchange_create_ack_record(rec)
+            self.with_delay(self._get_job_delay_params(rec)).exchange_create_ack_record(
```
self.with_delay(**self._get_job_delay_params(rec)).exchange_create_ack_record(rec)

etobella reviewed (commented):
I think there is some errors with the queue_job can you check?

simahawk commented:
> I think there is some errors with the queue_job can you check?

ehm, I was planning to add tests... I finished too late to add them :sweat_smile: 
I'll add tests anyway later.
Good catch mate, thanks :)

simahawk commented:
meanwhile, could you have a look at this if you have time https://github.com/OCA/edi/pull/447 ?

simahawk committed:
edi_backend: exclude class name from error txt

simahawk committed:
edi: exclude exception class from error msg

simahawk committed:
edi: improve exchange type and record UI

simahawk committed:
edi: add 'retry' action to exchange record

This action allows to re-schedule the record for the action that failed
depending on its current state and direction.

simahawk commented:
@etobella fixed and improved tests

simahawk commented:
@LoisRForgeFlow would you mind having a look here?

simahawk commented on edi_webservice_oca/views/edi_backend.xml:
```diff
@@ -6,11 +6,13 @@
-                </group>
-            </group>
+            <notebook position="inside">
+                <page name="webservice_config">
```
```suggestion
                <page name="webservice_config" string="Webservice">
```

LoisRForgeFlow commented on edi_oca/models/edi_exchange_record.py:
```diff
@@ -244,6 +260,18 @@ def action_exchange_process(self):
         self.ensure_one()
         return self.backend_id.exchange_process(self)
 
+    def action_retry(self):
+        if not self.retryable:
+            return False
+        new_state = self._rollback_state_mapping[self.edi_exchange_state]
+        fname = "edi_exchange_state"
+        self[fname] = new_state
+        display_state = self._fields[fname].convert_to_export(self[fname], self)
+        self.message_post(
+            body=_("Action retry: state moved back to '%s'") % display_state
+        )
+        return True
```
With this new action, I think we should not process failed files by default here: https://github.com/OCA/edi/blob/14.0/edi_oca/models/edi_backend.py#L543.

Also, it would be good if it accepts recordsets, to be potentially used in batch with a server action.

simahawk replied:
> With this new action, I think we should not process failed files by default here: https://github.com/OCA/edi/blob/14.0/edi_oca/models/edi_backend.py#L543.

I don't get this, can you elaborate? From my POV if the file failed either the system or the user can try it again. What is probably desiderable is that that system stops after N attempts to try retry them?

> Also, it would be good if it accepts recordsets, to be potentially used in batch with a server actio

Sure, I'll change it.

simahawk replied:
added loop for recordsets.

For the other issue I think we can discuss it on a 2nd round.

LoisRForgeFlow reviewed (commented):
Now that you are doing some improvements, let me bring up one topic: what do you think about accepting multiple records linked to a single edi exchange record?

simahawk commented:
> Now that you are doing some improvements, let me bring up one topic: what do you think about accepting multiple records linked to a single edi exchange record?

@LoisRForgeFlow ahh... just read your comment after our conversation at OCA booth :smile: 

We can write an RFC based on what we said (having not file content and lookup on the parent).

etobella commented on edi_oca/models/edi_exchange_record.py:
```diff
@@ -237,12 +260,35 @@ def action_exchange_process(self):
+        self.message_post(
+            body=_("Action retry: state moved back to '%s'") % display_state
+        )
+        return True
```
That will not work if exchange_file_auto_generate is not checked. As this record will not be processed by the cron. So, at the end, we need to add similar button that I proposed (with less logic, obviously) or add the logic here (I think it is better with the buttons)

etobella replied:
Not blocking, we can proceed with the merge and then check how to improve this :wink:

simahawk committed:
edi_webservice_oca: add missing page label

simahawk commented:
/ocabot merge minor
